File: rust/src/solutions/reverse_integer.rs
// LeetCode 7. Reverse Integer in Rust

use super::Solution;

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let (mut x, mut ans) = (x, 0);
        while x != 0 {
            if ans > std::i32::MAX / 10 || ans < std::i32::MIN / 10 {
                return 0;
            }
            ans *= 10;
            ans += x % 10;
            x /= 10;
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let x = 123;
        let ans = 321;
        assert_eq!(Solution::reverse(x), ans);
    }

    #[test]
    fn example2() {
        let x = -123;
        let ans = -321;
        assert_eq!(Solution::reverse(x), ans);
    }

    #[test]
    fn example3() {
        let x = 120;
        let ans = 21;
        assert_eq!(Solution::reverse(x), ans);
    }

    #[test]
    fn example4() {
        let x = 0;
        let ans = 0;
        assert_eq!(Solution::reverse(x), ans);
    }
}

File: cpp/src/solutions/reverse_integer.cpp
// LeetCode 7. Reverse Integer

#include "cpp/src/common/common.h"

class Solution {
 public:
  int reverse(int x) {
    int ans = 0;
    while (x != 0) {
      if (ans > numeric_limits<int>::max() / 10 ||
          ans < numeric_limits<int>::min() / 10)
        return 0;
      ans *= 10;
      ans += x % 10;
      x /= 10;
    }
    return ans;
  }
};

TEST(reverse, example1) {
  int x = 123, ans = 321;
  EXPECT_EQ(Solution().reverse(x), ans);
}

TEST(reverse, example2) {
  int x = -123, ans = -321;
  EXPECT_EQ(Solution().reverse(x), ans);
}

TEST(reverse, example3) {
  int x = 120, ans = 21;
  EXPECT_EQ(Solution().reverse(x), ans);
}

TEST(reverse, example4) {
  int x = 0, ans = 0;
  EXPECT_EQ(Solution().reverse(x), ans);
}